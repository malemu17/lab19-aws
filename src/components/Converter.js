import React from 'react';
import '../styles/styles.scss';


/**
 * class represent converter
 * @constructor  props 
 * set state for x, y and rate fields 
 */

class Converter extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            x: this.props.x ? this.props.x : '',
            y: this.props.y ? this.props.y : '',
            rate: this.props.rate ? this.props.rate : 1,
        };
    }
/**
 * this function round the decimal numbers 
 * @param {int} num 
 * @return rounded number to the nearest value 
 */
    roundIt(num) {
        return Math.round(num * 1000) / 1000;
    }

    /**
     * this function handles state 
     * @param {object} e 
     * @return  set empty/null if not object target value
     * @return  set value and roundvalue to  x or y  respectively 
     
     */

    handleChange(e) {
        if (!e.target.value) {
            this.setState({ x: '', y: '' });
            return;
        }

        let value = parseFloat(e.target.value);
        if (e.target.id === 'x')
            this.setState({
                x: value,
                y: this.roundIt(value * this.state.rate),
            });
        else if (e.target.id === 'y')
            this.setState({
                x: this.roundIt(value / this.state.rate),
                y: value,
            });
    }

    /**
     * render component converter , and all elemnts inside 
     * x & y labels
     * div element 
     * 
     */
    render() {
        let xLabel = this.props.xLabel ? this.props.xLabel : '1';
        let yLabel = this.props.yLabel || '1';
        return (
            <div className = "container">
            <div className='converter'>
                <h4 className='converter-heading'>
                    Converting {xLabel} to/from {yLabel}
                </h4>

                <label className='x-label'>
                    <input
                        type='number'
                        step='any'
                        id='x'
                        value={this.state.x}
                        onChange={this.handleChange.bind(this)}
                    />
                    {xLabel}
                </label>
                <span className='mid-label'> is equal to </span>
                <label className='y-label'>
                    <input
                        type='number'
                        step='any'
                        id='y'
                        value={this.state.y}
                        onChange={this.handleChange.bind(this)}
                    />

                    {yLabel}
                </label>
            </div>
            </div>
        );
    }
}

export default Converter;
