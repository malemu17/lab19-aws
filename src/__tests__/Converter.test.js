
import React from 'react';


import Converter from '../components/Converter';


import renderer from 'react-test-renderer';

// Mount (full, shallow), Render, testing
import { configure, mount, shallow, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe ('Converter correctly renders', ()=>{

  
  it ('Snapshot test', ()=>{

    const snap = renderer.create(<Converter />).toJSON();
    expect(snap).toMatchSnapshot();
  });

  it ('Test that changing the x input field properly changes the state', ()=>{
    let Component = mount(<Converter />);
    let label = Component.find('.x-label #x').at(0);
    label.simulate('change');
    expect(Component.state('rate')).toBe(1)
  });

  it ('Test that changing the y input field properly changes the state', ()=>{
    let Component = mount(<Converter />);
    let input = Component.find('.y-label #y').at(0);
    input.simulate('change');
    expect(Component.state('rate')).toEqual(1)
  });
    
 it ('Test that the component renders two input fields', ()=>{
   
   let component = render(<Converter />);
   expect(component.find('.x-label .y-label')).toBeDefined();
 });








it ('Test that the labels are correctly set', ()=>{
  let Component = shallow(<Converter />);
  expect(Component.state('x')).toBeDefined();
  expect(Component.state('y')).toBeDefined();

})


})