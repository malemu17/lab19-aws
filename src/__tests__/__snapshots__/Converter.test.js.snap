// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Converter correctly renders Snapshot test 1`] = `
<div
  className="container"
>
  <div
    className="converter"
  >
    <h4
      className="converter-heading"
    >
      Converting 
      1
       to/from 
      1
    </h4>
    <label
      className="x-label"
    >
      <input
        id="x"
        onChange={[Function]}
        step="any"
        type="number"
        value=""
      />
      1
    </label>
    <span
      className="mid-label"
    >
       is equal to 
    </span>
    <label
      className="y-label"
    >
      <input
        id="y"
        onChange={[Function]}
        step="any"
        type="number"
        value=""
      />
      1
    </label>
  </div>
</div>
`;
